### PR TITLE
Add Aumann-Shapley AutoQuantize recipe integration

### PR DESCRIPTION
**Pull request by joshua-hill**
[Paper](https://arxiv.org/abs/2607.12266) · [Overview](https://x.com/waterloo_intern/status/2076460984475263401) · [Implementation thread](https://x.com/the_joshua_hill/status/2076427869388255635)

Depends on #2183, which adds the Aumann-Shapley AutoQuantize method, and transitively on #2231. Until #2183 merges, GitHub's default diff also shows the parent commits; the diff against these 2 prior PRs are [here](https://github.com/joshua-hill/Model-Optimizer/compare/feat/aumann-shapley-autoquant...feat/aumann-shapley-recipe-integration).

### What does this PR do?

Type of change: new feature

This PR makes the Aumann-Shapley AutoQuantize method available through ModelOpt recipes and the Hugging Face PTQ example.

An AutoQuantize recipe can now select `auto_quantize_method: aumann_shapley` and pass method-specific settings through `method_options`. The recipe can choose one of two search targets:

- an `effective_bits` target, which selects the lowest-damage configuration within a bit budget; or
- `max_predicted_damage`, which selects the lowest-cost configuration within a predicted-damage budget.

The recipe schema rejects configurations that specify both targets, and it requires the Aumann-Shapley method when `max_predicted_damage` is used. Detailed method-option validation remains in `mtq.auto_quantize`, so the core API stays the single source of truth.

The `hf_ptq` integration forwards the method options, uses the existing label-free logits path, and removes the schema's default bit target when the recipe selects a predicted-damage target. Existing gradient and KL-divergence recipe behavior is unchanged.

The Hugging Face PTQ README documents both recipe forms in plain YAML.

### Usage

Target an effective bit width:

```yaml
auto_quantize:
  constraints:
    effective_bits: 5.4
  auto_quantize_method: aumann_shapley
  method_options:
    num_path_nodes: 2
    damage_link: coverage
```

Or target predicted damage:

```yaml
auto_quantize:
  constraints: {}
  auto_quantize_method: aumann_shapley
  method_options:
    num_path_nodes: 2
    max_predicted_damage: 0.01
```

These fragments fit into the existing AutoQuantize recipe format alongside `candidate_formats`, `score_size`, and the existing layer-selection fields.

### Testing

```text
pytest -o addopts='' \
  tests/unit/recipe/test_loader.py \
  tests/unit/recipe/test_recipe_docs.py \
  tests/examples/hf_ptq/test_hf_ptq_args.py \
  tests/unit/torch/quantization/test_autoquant.py \
  tests/unit/torch/quantization/test_autoquant_shapley.py
```

Result: **473 passed**.

All pre-commit hooks pass on the five changed files.

Recipe-driven GPU smoke tests also passed on the locally cached `Qwen/Qwen2.5-0.5B-Instruct` model with NVFP4 and FP8 candidates:

- The effective-bits recipe completed at 5.40 effective bits and produced finite logits.
- The predicted-damage recipe reported a satisfied constraint, selected a 4.56-effective-bit configuration, and produced finite logits.

An additional recipe-driven smoke test passed on `Qwen/Qwen3-30B-A3B` (128 experts, 8 active experts per token): 145 module groups were scored, the solver selected a mixed FP8/BF16 recipe, the damage model and constraint were valid, and the post-quantization logits were finite.

### Before your PR is "Ready for review"

- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow the contributing guidance?: ✅ No copied code and no new dependencies.
- Did you write the necessary tests?: ✅
- Did you update `CHANGELOG.rst`?: N/A — the underlying feature entry is included in #2183.
- Are the commits signed and signed off?: ✅
- Did you get Claude approval on this PR?: N/A

## Summary by CodeRabbit

* **New Features**
  * Added the label-free `aumann_shapley` method to automatic quantization.
  * Added configurable scoring options, predicted-damage bounds, and minimum-cost optimization.
  * Added additive and coverage damage models, checkpoint resumption, and diagnostic scoring metadata.

* **Bug Fixes**
  * Improved score handling, distributed processing, configuration validation, and recovery from scoring or setup failures.

* **Documentation**
  * Expanded the post-training quantization guide with Aumann–Shapley configuration, constraints, and backpropagation requirements.

### CODE REVIEW

File: modelopt/torch/quantization/algorithms.py [hunk 1]
```diff
@@ -1434,11 +1504,252 @@ def _get_auto_quantize_score(grad_output, output_diff):
...
+            self._replay_candidates(module, base, lambda hparam: hparam.choices, *args, **kwargs)
+        )
+        return output
+
+    def backward_hook(self, module: nn.Module, grad_input, grad_output) -> None:
+        """Accumulate squared gradient-weighted output perturbations."""
+        invocation_diffs = self._output_diffs[module].pop()
+        if not self._output_diffs[module]:
+            del self._output_diffs[module]
+        self._accumulate_candidate_scores(module, invocation_diffs, grad_output[0])
```
[the remark this answers is not in the document]
Comment (joshua-hill) [reply modelopt/torch/quantization/algorithms.py]: Addressed in 14b2af2 and propagated here in dd97561. Gradient replay now returns early when base.requires_grad is false, and the reused-module regression includes a detached invocation with no matching backward hook. The final 473-test stack passes.

File: modelopt/torch/quantization/algorithms.py [hunk 2]
```diff
@@ -1434,11 +1504,252 @@ def _get_auto_quantize_score(grad_output, output_diff):
...
```
[the remark this answers is not in the document]
Comment (joshua-hill) [reply modelopt/torch/quantization/algorithms.py]: Addressed in 00aba57 and propagated through 2683617 and 0287875. The module-wide replay stack and full backward hook have been removed: each differentiable output now owns a gradient hook that closes over that invocation’s replay differences. An unused differentiable output therefore cannot leave stale state or shift attribution to another call. The gradient and Aumann–Shapley reused-module regressions cover unused invocations; the 473-test stack and Qwen3-30B-A3B MoE smoke both pass.
